pilot/pkg/serviceregistry/kube/controller/endpointslice.go

```diff
@@ -0,0 +1,292 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"fmt"
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	discoveryv1alpha1 "k8s.io/api/discovery/v1alpha1"
+	klabels "k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/serviceregistry/kube"
+	"istio.io/istio/pkg/config/host"
+	configKube "istio.io/istio/pkg/config/kube"
+	"istio.io/istio/pkg/config/labels"
+	"istio.io/pkg/log"
+)
+
+type endpointSliceController struct {
+	cache         cacheHandler
+	endpointCache *endpointSliceCache
+}
+
+var _ edsController = &endpointSliceController{}
+
+func newEndpointSliceController(c *Controller, sharedInformers informers.SharedInformerFactory) *endpointSliceController {
+	epSliceInformer := sharedInformers.Discovery().V1alpha1().EndpointSlices().Informer()
+	// TODO Endpoints has a special cache, to filter out irrelevant updates to kube-system
+	// Investigate if we need this, or if EndpointSlice is makes this not relevant
+	return &endpointSliceController{
+		cache:         c.createCacheHandler(epSliceInformer, "EndpointSlice"),
+		endpointCache: newendpointSliceCache(),
+	}
+}
+
+func (e endpointSliceController) Get() cacheHandler {
+	return e.cache
+}
+
+func (e endpointSliceController) updateEDSSlice(c *Controller, slice *discoveryv1alpha1.EndpointSlice, event model.Event) {
+	svcName := slice.Labels[discoveryv1alpha1.LabelServiceName]
+	hostname := kube.ServiceHostname(svcName, slice.Namespace, c.domainSuffix)
+
+	endpoints := make([]*model.IstioEndpoint, 0)
+	if event != model.EventDelete {
+		for _, e := range slice.Endpoints {
+			if e.Conditions.Ready != nil && !*e.Conditions.Ready {
+				// Ignore not ready endpoints
+				continue
+			}
+			for _, a := range e.Addresses {
+				pod := c.pods.getPodByIP(a)
+				if pod == nil {
+					// This can not happen in usual case
+					if e.TargetRef != nil && e.TargetRef.Kind == "Pod" {
+						log.Warnf("Endpoint without pod %s %s.%s", a, svcName, slice.Namespace)
+
+						if c.metrics != nil {
+							c.metrics.AddMetric(model.EndpointNoPod, string(hostname), nil, a)
+						}
+						// TODO: keep them in a list, and check when pod events happen !
+						continue
+					}
+					// For service without selector, maybe there are no related pods
+				}
+
+				var labels map[string]string
+				locality := GetLocalityFromTopology(e.Topology)
+				sa, uid := "", ""
+				if pod != nil {
+					sa = kube.SecureNamingSAN(pod)
+					uid = fmt.Sprintf("kubernetes://%s.%s", pod.Name, pod.Namespace)
```

Comment:
nit: same comment wrt perf ... should probably make this a utility and change all the call sites.

Reply:
done, broke out to method